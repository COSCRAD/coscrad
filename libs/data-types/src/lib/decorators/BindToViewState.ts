import { COSCRAD_DATA_TYPE_METADATA } from '../constants';
import getCoscradDataSchemaFromPrototype from '../utilities/getCoscradDataSchemaFromPrototype';

export function BindToViewState(viewPropertyName: string): PropertyDecorator {
    return (target: Object, propertyKey: string) => {
        const existingMeta = getCoscradDataSchemaFromPrototype(target);

        const existingMetaForThisProp = existingMeta[propertyKey] || {};

        Reflect.defineMetadata(
            COSCRAD_DATA_TYPE_METADATA,
            {
                ...existingMeta,
                [propertyKey]: {
                    ...existingMetaForThisProp,
                    bindToViewProperty: viewPropertyName,
                },
            },
            target
        );
    };
}
