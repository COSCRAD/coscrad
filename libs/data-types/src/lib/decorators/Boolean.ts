import { CoscradDataType } from '../types';
import appendMetadata from '../utilities/appendMetadata';
import mixinDefaultTypeDecoratorOptions from './common/mixinDefaultTypeDecoratorOptions';
import { TypeDecoratorOptions } from './types';

// we have named this deorator to avoid collisions with the wrapper class Boolean
export function BooleanDataType(userOptions: TypeDecoratorOptions): PropertyDecorator {
    return (target: Object, propertyKey: string | symbol) => {
        const options = mixinDefaultTypeDecoratorOptions(userOptions);

        appendMetadata(target, propertyKey, CoscradDataType.Boolean, options);
    };
}
