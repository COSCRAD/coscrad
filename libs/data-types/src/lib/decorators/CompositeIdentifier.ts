import { CoscradDataType } from '../types';

import appendMetadata from '../utilities/appendMetadata';
import mixinDefaultTypeDecoratorOptions from './common/mixinDefaultTypeDecoratorOptions';
import { TypeDecoratorOptions } from './types/TypeDecoratorOptions';

// TODO move this to a utility types lib
type TypeGuard<T> = (input: unknown) => input is T;

// @deprecated Use @NestedDataType(CustomCompositeIdentifierClass,options) instead
export function CompositeIdentifier(
    AllowedTypesEnum: Record<string, string>,
    idTypeGuard: TypeGuard<string | number>,
    userOptions: TypeDecoratorOptions
): PropertyDecorator {
    return (target: Object, propertyKey: string | symbol) => {
        const options = { ...mixinDefaultTypeDecoratorOptions(userOptions) };

        appendMetadata(target, propertyKey, CoscradDataType.CompositeIdentifier, options);
    };
}
