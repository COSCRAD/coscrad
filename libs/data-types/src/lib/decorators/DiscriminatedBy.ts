import { COSCRAD_UNION_MEMBER_METADATA } from '../constants';

/**
 * This decorator factory function allows you to register the value of the
 * discriminant that distinguishes it from other members of a discriminated union.
 * This will only take effect if the target class is referenced in the
 * `@Union(...)` decorator, and should be the value of the property specified
 * there via the `discriminantPath`.
 */
export function DiscriminatedBy(discriminantValue: string): ClassDecorator {
    return (target: Object): void => {
        Reflect.defineMetadata(
            COSCRAD_UNION_MEMBER_METADATA,
            {
                discriminantValue,
            },
            target
        );
    };
}
