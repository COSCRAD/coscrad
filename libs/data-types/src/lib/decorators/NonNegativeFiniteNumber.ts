import CoscradDataType from '../types/CoscradDataType';
import appendMetadata from '../utilities/appendMetadata';
import buildDefaultTypeDecoratorOptions from './common/buildDefaultTypeDecoratorOptions';
import { TypeDecoratorOptions } from './types/TypeDecoratorOptions';

export function NonNegativeFiniteNumber(
    options: TypeDecoratorOptions = buildDefaultTypeDecoratorOptions()
): PropertyDecorator {
    return (target: Object, propertyKey: string | symbol) => {
        appendMetadata(target, propertyKey, CoscradDataType.NonNegativeFiniteNumber, options);
    };
}
