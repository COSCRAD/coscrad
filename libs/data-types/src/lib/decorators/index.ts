export * from './CompositeIdentifier';
export * from './Enum';
export * from './NestedDataType';
export * from './NonEmptyString';
export * from './NonNegativeFiniteNumber';
export * from './RawData';
export * from './URL';
export * from './UUID';
