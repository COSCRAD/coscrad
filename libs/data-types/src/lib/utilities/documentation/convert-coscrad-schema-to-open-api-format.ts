import { isNullOrUndefined } from '@coscrad/validation-constraints';
import { SchemaObject } from '@nestjs/swagger/dist/interfaces/open-api-spec.interface';
import {
    ClassSchema,
    CoscradDataType,
    CoscradPropertyTypeDefinition,
    isSimpleCoscradPropertyTypeDefinition,
} from '../../types';

export const convertCoscradSchemaToOpenApiFormat = (
    schema: ClassSchema<Record<string, unknown>>
): SchemaObject => {
    if (isNullOrUndefined(schema)) {
        return {};
    }

    const properties = Object.entries(schema).reduce(
        (acc, [propertyName, typeDefinition]: [string, CoscradPropertyTypeDefinition]) => {
            if (!isSimpleCoscradPropertyTypeDefinition(typeDefinition)) {
                return acc;
                // throw new Error(
                //     `Unsupported property schema (only simple types are supported): ${JSON.stringify(
                //         typeDefinition
                //     )}`
                // );
            }

            const openApiSchemaForThisProperty: SchemaObject = {
                type:
                    typeDefinition.coscradDataType === CoscradDataType.NonNegativeFiniteNumber
                        ? 'number'
                        : typeDefinition.coscradDataType,
                description: 'non-negative, finite number',
                minimum: 0,
            };

            acc[propertyName] = openApiSchemaForThisProperty;

            return acc;
        },
        {}
    );

    return {
        properties,
        description: 'we should put a description here',
    };
};
