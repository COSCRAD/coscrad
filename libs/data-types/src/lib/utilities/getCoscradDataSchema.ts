import { ClassSchema } from '../types/ClassSchema';
import getCoscradDataSchemaFromPrototype from './getCoscradDataSchemaFromPrototype';

// eslint-disable-next-line
export default (TargetClass: Object): ClassSchema<Record<string, unknown>> =>
    // @ts-expect-error TODO: restrict argument to be a class
    getCoscradDataSchemaFromPrototype(TargetClass.prototype || {});
