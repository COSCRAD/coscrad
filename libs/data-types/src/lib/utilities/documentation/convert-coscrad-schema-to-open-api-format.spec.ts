import { NonNegativeFiniteNumber } from '../../decorators/NonNegativeFiniteNumber';
import getCoscradDataSchema from '../getCoscradDataSchema';
import { convertCoscradSchemaToOpenApiFormat } from './convert-coscrad-schema-to-open-api-format';

describe('convertCoscradSchemaToOpenApiFormat', () => {
    describe(`when there is a simple schmea with only basic data types`, () => {
        it(`should return the correct open API schema`, () => {
            class Simple {
                @NonNegativeFiniteNumber({
                    label: 'count',
                    description: 'how many are there?',
                })
                count: number;
            }

            const openApiSchema = convertCoscradSchemaToOpenApiFormat(getCoscradDataSchema(Simple));

            expect(openApiSchema).toMatchSnapshot();
        });
    });
});
