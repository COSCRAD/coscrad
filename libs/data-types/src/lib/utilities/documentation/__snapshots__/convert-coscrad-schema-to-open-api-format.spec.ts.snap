// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertCoscradSchemaToOpenApiFormat when there is a simple schmea with only basic data types should return the correct open API schema 1`] = `
{
  "properties": {
    "count": {
      "type": "NON_NEGATIVE_FINITE_NUMBER",
    },
  },
}
`;
